metadata/src/utils/retain.rs

```diff
@@ -261,13 +314,37 @@ pub fn retain_metadata<F, G>(
         .iter()
         .find(|ty| ty.ty.path.segments == ["sp_runtime", "DispatchError"])
         .expect("Metadata must contain sp_runtime::DispatchError");
-    type_ids.insert(dispatch_error_ty.id);
+    type_set.seen_ids.insert(dispatch_error_ty.id);
+    type_set.seen_ids.insert(metadata.runtime_ty);
+    type_set.collect_extrinsic_types(&metadata.extrinsic);
+
+    // Collect the outer enums type IDs.
+    for typ in [
+        metadata.outer_enums.call_enum_ty,
+        metadata.outer_enums.event_enum_ty,
+        metadata.outer_enums.error_enum_ty,
+    ] {
+        type_set.collect_variants_in_type(metadata, typ, &mut pallets_filter);
+    }
+
+    // Filter out unnecesary pallets that have no entries
+    metadata
+        .pallets
+        .retain(|pallet| pallets_filter(&pallet.name));
 
-    // Now, keep the type IDs we've asked for. This recursively keeps any types referenced from these.
-    // This will return a map from old to new type ID, because IDs may change.
-    let map_ids = metadata.types.retain(|id| type_ids.contains(&id));
+    // Retain the apis
+    metadata.apis.retain(|api| runtime_apis_filter(&api.name));
+
+    metadata.pallets_by_index = metadata
```

Comment:
this is a bit confusing to my eyes, but I guess you are updating the pallet_index after possible removing pallets via retain above?

Reply:
yup, added a comment